### ISSUE TITLE
Semantics of Method Invocation

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/ModelicaSpecification#59]
**Reported by mtiller on 11 Mar 2008 17:31 UTC**
For some time I have wanted to have methods in Modelica.  This was considered an extremely advanced topic and therefore there has not been much progress on it.

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/59

**Comment by modelica-trac-importer** [comment]
**Comment by mtiller on 11 Mar 2008 17:58 UTC**
During the 56th design meeting, I thought about this a bit because of various projects I'm involved in.  My feeling is that it should be possible to make some minor adjustments to the lookup rules and achieve most of what I need.  First, it is necessary to ease one restriction imposed in Section 5.3.1 in the Modelica 3.0 specification to read as follows:

  Reference to variables successfully looked up in an enclosing class are only allowed for a) variables declared as constants **or b) variables declared in the immediately enclosing scope if the lookup started within a method**. If the use is in a modifier on a short class definition, see Section 4.5.1.

One additional change to Section 5.3.2 is also required:

For a composite name of the form A.B [or A.B.C, etc.] lookup is performed as follows:
  * The first identifier [A] is looked up as defined above.
  * If the first identifier denotes a component, lookup of the rest of the name [e.g., B or B.C] is performed as follows:
     * The next identifier is looked up among the declared named component elements of the component.
     * **The identifier is checked against method names defined within the class definition of the component**.
  * If the identifier denotes a class, that class is temporarily flattened with an empty environment (i.e. no modifiers, see Section 7.2.2) and using the enclosing classes of the denoted class. The rest of the name [e.g., B or B.C] is looked up among the declared named elements of the temporary flattened class. If the class does not satisfy the requirements for a package, the lookup is restricted to encapsulated elements only. The class we look inside may not be partial in a simulation model.

**Comment by modelica-trac-importer** [comment]
**Comment by mtiller on 11 Mar 2008 19:50 UTC**
I've added a new Wiki page MethodsMotiviation, to discuss the motivations behind wanting to include methods in the language.

**Comment by modelica-trac-importer** [comment]
**Comment by mtiller on 11 Mar 2008 20:03 UTC**
Replying to [comment:2 mtiller]:
Oops, Trac didn't include the link.  The page with the motivation for methods on it is [wiki:MethodsMotivation].

**Comment by modelica-trac-importer** [comment]
**Comment by dietmarw on 11 Mar 2008 22:28 UTC**
Sorry for being pedantic but MSL3.1 is hardly a matching Milestone for an --ModelicaSpecification-- issue ;-)

**Comment by modelica-trac-importer** [comment]
**Comment by dietmarw on 3 Jul 2008 16:21 UTC**
The page with the motivation for the methods was moved to wiki:Documents/MethodsMotivation

**Comment by modelica-trac-importer** [comment]
**Modified by otter on 15 Aug 2008 17:25 UTC**

**Comment by modelica-trac-importer** [comment]
**Modified by dietmarw on 11 Sep 2008 08:32 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by magnus on 18 Nov 2008 15:51 UTC**

We currently use variants of the following construct to get something like immutable methods:

```
function Functions.f
  input Real r;
  ...
end Functions.f;

model M
  ...
  Real r;
  function f = Functions.f(r=r);
end M;
```

I guess the proposal means to introduce something like
```
function Functions.f1
  ...
end Functions.f1;

model M
  ...
  Real r;
  method f = Functions.f1;
end M;
```
where r would be accessible from within the function, without have it declared as an input in the function.

**Comment by modelica-trac-importer** [comment]
**Comment by mtiller on 18 Nov 2008 21:58 UTC**
The earlier link (wiki:Documents/MethodsMotivation) was primarily to document why I think methods will be important in Modelica.  I've refined my ideas after discussions with Martin and Sebastien and borrowing some ideas from Peter Fritzson.  The result is something that is a much more concrete proposal and it can be found at  wiki:Documents/FunctionInstantiation.

Note that it does not completely address the issue of methods.  Sebastien correctly pointed out that to have true methods it would be necessary to include some additional semantics but what I have proposed could be taken as a first step with additional semantics added later.

I am quite happy with this approach since I think it provides a very simple and useful functionality that is fairly easy to read and understand.

I had several "side" discussions about this proposal and I thought I had submitted it to the overall group but it seems perhaps I didn't.  Sorry about the confusion.

**Comment by modelica-trac-importer** [comment]
**Comment by mtiller on 18 Nov 2008 22:12 UTC**
Magnus,

  To directly address your comment.  My proposal would turn your example code into this:

```text/x-modelica
partial function RoadSurface
  input Real x;
  input Real y;
  output Real z;
end RoadSurface;

function Functions.f1
  extends RoadSurface;
  input Real r;
algorithm
  z := f(r, x, y);
end Functions.f1;

model M
  ...
  closure RoadSurface surface = Functions.f1(r=r);
end M;
```

However, a more "likely" scenario would be something like this:

```text/x-modelica
partial function RoadSurface
  input Real x;
  input Real y;
  output Real z;
end RoadSurface;

function TableBasedSurface
  extends RoadSurface;
  input TableData data;
algorithm
  z := f(data, x, y);
end TableBasedData;

model Tire
  ...
  closure RoadSurface surface "No value, to be added later";
end Tire;

partial model VehicleArchitecture
  closure RoadSurface surface "Again, no value but propagated";
  Tire left_front(surface=surface);
  Tire right_front(surface=surface);
  Tire left_rear(surface=surface);
  Tire right_rear(surface=surface);
  ...
end VehicleArchitecture;

model MyVehicleDesign
  extends Vehicle(surface=TableBasedSurface(data=RoadSurface.TestTrack));
end MyVehicleDesign;

model FlatWorldExperiment
  extends MyVehicleDesign(surface=FlatSurface(elevation=25));
end FlatWorldExperiment;
```

The point is that "surface" is a first class object that can be propagated just like a value (made a parameter, used with inner/outer, pass as an argument to a function, as an input to a model, etc) but it is invoked like a function (i.e. function as a first class object).  See the wiki page for how this address three important cases (including methods).

**Comment by modelica-trac-importer** [comment]
**Comment by mtiller on 18 Dec 2008 14:24 UTC**
It is worth pointing out that the proposals that I am putting forward do not use the redeclaration mechanism in Modelica to achieve "passing" of functions.  The main reason is that redeclarations apply to types, not values.  I feel it is essential for all the use cases demonstrated in wiki:Documents/FunctionInstantiation that functions can be represented as values.  Although redeclaration works when passing functions to functions, it is much less practical for things like propagation of property evaluators through connections and the implementation of methods (or at least method-like) functions.

So, while utilizing redeclarations may require fewer modifications to the specification, it is (in my opinion) more awkward and less practical for the majority of use cases.

**Comment by modelica-trac-importer** [comment]
**Comment by dietmarw on 20 Feb 2009 15:46 UTC**
closing milestone:Design60

**Comment by modelica-trac-importer** [comment]
**Comment by HansOlsson on 17 Mar 2009 12:28 UTC**
Proposal text complete.

**Comment by modelica-trac-importer** [comment]
**Comment by dietmarw on 11 Jun 2009 18:45 UTC**
Automatically pushed to milestone:Design63 due to closing of milestone:Design62.

**Comment by modelica-trac-importer** [comment]
**Modified by dietmarw on 19 Nov 2009 16:36 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by choeger on 9 Jun 2010 10:43 UTC**
Please note that Modelica methods take values as input, whereas methods would take reference parameters. Using the same construct for both concepts would be misleading.

The natural way of doing this in Modelica would be:

```
function Functions.f
  outer Real r;
  ...
end Functions.f;

model M
  ...
  inner Real r;
  function f = Functions.f();
end M;
```

**Comment by modelica-trac-importer** [comment]
**Comment by HansOlsson on 15 Sep 2010 15:07 UTC**
Planned to include member functions in some way in Modelica 4.

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 21 Mar 2018 17:52 UTC**
Seems that most of this was corrected in Modelica 3.3 with 12.4.2.1 function partial application.

**Comment by modelica-trac-importer** [comment]
**Changelog removed by hansolsson on 21 Mar 2018 17:52 UTC**
